components/places/src/ffi.rs

```diff
@@ -43,6 +48,155 @@ impl UniffiCustomTypeWrapper for Url {
+    /// provide an earlier (but still valid) timestamp.
+    ///
+    /// - Always returned on reads.
+    /// - Ignored for insertion and update.
+    pub date_added: Option<i64>,
+    ///*
+    /// Last modification time, in milliseconds since the unix epoch.
+    ///
+    /// - Always returned on reads.
+    /// - Ignored for insertion and update.
+    pub last_modified: Option<i64>,
+    ///*
+    /// Guid of the parent record.
+    ///
+    /// - Returned on reads, except for reads of the bookmark root.
+    /// - Required for insertion.
+    /// - On updates, if provided, we treat it as a move.
+    ///     - Interacts with `position`, see its documentation below
+    ///       for details on how.
+    pub parent_guid: Option<String>,
+    ///*
+    /// Zero based index within the parent.
+    ///
+    /// - Not provided on reads (for now).
+    ///
+    /// - Allowed for insertion.
+    ///    - Leaving it out means 'end of folder'.
+    ///
+    /// - Allowed for updates.
+    ///     - If `parent_guid` is not provided and `position` is, we treat this
+    ///       a move within the same folder.
+    ///
+    ///     - If `parent_guid` and `position` are both provided, we treat this as
+    ///       a move to / within that folder, and we insert at the requested
+    ///       position.
+    ///
+    ///     - If `position` is not provided (and `parent_guid` is) then it's
+    ///       treated as a move to the end of that folder.
+    pub position: Option<u32>,
+    ///*
+    /// Bookmark title. Not present for type = `BookmarkType::Separator`.
+    ///
+    /// - Returned on reads if it exists.
+    /// - Required when inserting folders.
+    pub title: Option<String>,
+    ///*
+    /// Bookmark URL. Only allowed/present for type = `BookmarkType::Bookmark`.
+    ///
+    /// - Always returned on reads (for `BookmarkType::Bookmark`).
+    /// - Required when inserting a new bookmark.
+    pub url: Option<String>,
+    ///*
+    /// IDs of folder children, in order. Only present for type =
+    /// `BookmarkType::Folder`.
+    ///
+    /// - Returned on reads (for `BookmarkType::Folder`).
+    /// - Forbidden for insertions and updates.
+    /// - Not provided if `child_nodes` is provided, to avoid sending more data
+    ///   over the FFI than necessary.
+    pub child_guids: Vec<String>,
+    ///*
+    /// Data about folder children, in order. Only present for type =
+    /// `BookmarkType::Folder`.
+    ///
+    /// For performance reasons, this only is provided if it's requested.
+    pub child_nodes: Vec<BookmarkNode>,
+    ///*
+    /// Returned by reads, and used to distinguish between the cases of
+    /// "empty child_nodes because the API doesn't return children" and
+    /// "empty child_nodes because this folder has no children (but
+    /// we'd populate them if it had them)".
+    ///
+    /// Only required because you can't have `optional repeated`.
```

Comment:
I think in this world we *could* have child_nodes as `Option<Vec<BookmarkNode>>` although we might decide that's beyond the scope of this PR - so maybe just update the comment to indicate that "when we used protobufs..."